python/vineyard/core/client.py

```diff
@@ -91,21 +100,47 @@ def __init__(
             socket = os.getenv('VINEYARD_IPC_SOCKET', None)
         if not endpoint and not (host and port):
             endpoint = os.getenv('VINEYARD_RPC_ENDPOINT', None)
-        if endpoint:
-            if not isinstance(endpoint, (tuple, list)):
-                endpoint = endpoint.split(':')
+        if isinstance(endpoint, Tuple[str, Union[str, int]]):
             host, port = endpoint
 
+        vineyard_config = {}
+        if config:
+            if os.path.isdir(config):
+                config = os.path.join(config, 'vineyard.yaml')
+
+            if os.path.isfile(config):
+                with open(config, 'r', encoding='utf-8') as f:
+                    vineyard_config = yaml.safe_load(f).get('Vineyard', {})
+
+        if not socket and vineyard_config:
+            socket_config = vineyard_config.get('IPCSocket', None)
+            if socket_config:
+                base_dir = os.path.dirname(config) if os.path.isfile(config) else config
+                socket = os.path.join(base_dir, socket_config)
+
+        if not endpoint and not (host and port) and vineyard_config:
+            endpoint = vineyard_config.get('RPCEndpoint', None)
+
         if socket:
             self._ipc_client = _connect(socket, **kwargs)
         if host and port:
             self._rpc_client = _connect(host, port, **kwargs)
+        # try to connect the first available rpc endpoint
+        if endpoint and not self._rpc_client:
+            for ep in endpoint.split(','):
+                try:
+                    host, port = [e.strip() for e in ep.split(':')]
+                    self._rpc_client = _connect(host, port, **kwargs)
+                    break
+                except Exception:
+                    warnings.warn(f"Failed to connect to {ep}, try next endpoint...")
+                    continue
 
         if self._ipc_client is None and self._rpc_client is None:
             raise ConnectionError(
                 "Failed to connect to vineyard via both IPC and RPC connection. "
-                "Arguments are not and environment variables VINEYARD_IPC_SOCKET "
-                "and VINEYARD_RPC_ENDPOINT are not available."
+                "Arguments and environment variables VINEYARD_IPC_SOCKET "
+                "and VINEYARD_RPC_ENDPOINT and config file are not available."
```

Comment:
You could let chatgpt help to prootread the documentation and warning messages for better writting.